src/main/kotlin/io/emeraldpay/dshackle/upstream/ethereum/EthereumMultistream.kt

```diff
@@ -142,6 +147,25 @@ open class EthereumMultistream(
         return subscribe
     }
 
+    override fun getHead(mather: Selector.Matcher): Head =
+        filteredHeads.computeIfAbsent(mather.describeInternal()) { _ ->
+            upstreams.filter { mather.matches(it) }
+                .apply {
+                    log.debug("Found $size upstreams matching [${mather.describeInternal()}]")
+                }
+                .map { it.getHead() }
+                .let {
+                    when (it.size) {
+                        0 -> EmptyHead()
+                        1 -> it.first()
+                        else -> MergedHead(it, MostWorkForkChoice()).apply {
+                            start()
+                        }
+                    }
+                }
+        }
+    // TODO track unused heads and remove
```

Comment:
should we do that now?

Reply:
fixed by using ConcurrentReferenceHashMap with WEAK reference option. This means that items may be removed from the map in order to free memory  - managed by JVM